### PR TITLE
fix(genai,#12428): FT-04 dataset preferences 126 paires + eval disjointe par fond — mémorisation DPO rendue visible

### LINKED ISSUE #12428: FineTuning FT-03/FT-04 — datasets dégénérés (20 et 6 paires) : charabia committé sans verdict, métrique de succès trompeuse
## Constat (outputs committés vérifiés)

Les deux notebooks centraux de la série `GenAI/FineTuning` entraînent sur des datasets si petits que les résultats committés sont du **charabia intégral — sans qu'aucune cellule n'arbitre ce que l'étudiant est censé en conclure**.

**FT-03-Supervised-FineTuning-SFT** :
- Dataset : **20 paires** instruction/réponse, entraînement **31.4 s**, loss finale 2.92.
- Sorties committées : « *A gravite, un objectif est établi par les images de l'objet* », « *Ne pas pousser la petite quantitatrice* », et la démo 5-exemples : « *La homhomig-----(,,,,,s. (e. ---/ in,"--im,,I(* ».
- Aucune cellule markdown ne dit : à 20 exemples le SFT ne peut pas marcher, voici pourquoi, voici ce qu'on observe quand même (format ### Assistant appris, perte qui descend). L'étudiant voit un fine-tuning « réussi » (le code tourne) dont la sortie est du non-sens, sans grille de lecture.

**FT-04-RLHF-DPO** :
- Dataset : **6 paires** de préférences ; SFT de référence entraîné sur **5 exemples** ; DPO : 3 epochs.
- Sortie finale : « Taux de preference correct (DPO) : 6/6 (100%) » — présenté pendant que les réponses générées restent du charabia (« *REST est une API, une API est un API de la langue Java* »).
- Le contraste « métrique parfaite / générations absurdes » est LA leçon la plus profonde du DPO à petite échelle (optimiser des log-probs sur 6 paires = mémorisation triviale, pas de l'alignement) — et le notebook ne la donne pas.

## Correction attendue

1. **Grossir les datasets** à la taille minimale où le phénomène démontré devient réel (pour FT-03 : quelques centaines de paires suffisent sur OPT-1.3B+QLoRA vu les 31 s actuelles ; pour FT-04 : ≥ 100 paires de préférences, réutilisation possible du dataset `train_prefs` 61k déjà chargé dans PT_02/PT_03).
2. **Arbitrer explicitement** : après chaque comparaison avant/après, une cellule markdown qui dit si le changement est attendu, surprenant, ou trivial à cette échelle. En particulier FT-04 doit commenter le 100%-sur-6-paires (mémorisation) vs un score sur éval disjointe.
3. **Éval disjointe** pour FT-04 (le taux d'apprentissage sur les paires d'entraînement ne prouve rien) — PT_03 a déjà la mécanique `preference accuracy` sur subset neutre, la réutiliser.
4. **Courbe de loss** (même simple, matplotlib) dans les deux notebooks : aucune courbe d'entraînement n'existe dans toute la série FT-01→FT-05.
5. Conserver ce qui marche : la progression pédagogique (log-probs → SFT → DPO), la mesure VRAM, FT-02 en contre-exemple positif (loss 2.41, VRAM pic 1.16 Go, 69.5 s — réel et mesuré).

## Critères d'acceptation

- [ ] FT-03 : ≥ 200 paires, loss curve committée, générations avant/après + verdict écrit
- [ ] FT-04 : ≥ 100 paires, éval disjointe rapportée (pas seulement le train set), verdict sur l'écart métrique/générations
- [ ] Zéro sortie charabia non commentée (chaque sortie dégradée porte son explication)
- [ ] Re-exécution complète (C.2), séquence CLEAN, exercices stubs préservés (C.1)

## Non-cible

FT-01 (7 segments comme première mécanique) et FT-05 (routeur 14 points) : traités séparément — leur problème est l'absence d'arbitrage statistique, pas des sorties absurdes.

### PR DESCRIPTION
**Pull request by jsboige**
Grain: DEEP/notebook-python — lane myia-po-2024:CoursIA — prev: DEEP/notebook-python #12594

## Résumé

Tranche FT-04 de l'issue #12428 : le dataset de préférences passe de **6 paires inline** (dont sortait le « Taux de preference correct : 6/6 (100%) » de mémorisation présenté pendant que les générations restaient du charabia) à un **builder déterministe de 126 paires**, et le DPO est instrumenté d'une **évaluation disjointe par fond** qui remplace la métrique trompeuse.

## Ce qui change

**Dataset (cellules 6-8)** — builder `_build_preference_dataset(seed=42)` : 42 fonds (20 définitions + 12 comparaisons + 10 pourquois, même corpus que FT-03) × **3 familles de rejets** = 126 paires. Les rejets sont construits, pas décorés à la main : **vague** (creux, aucun fait, même longueur que le choisi — le modèle ne peut pas apprendre « court = mauvais »), **expéditive** (familier, trivial), **hors-sujet** (réponse choisie d'un autre fond, verbatim). Déterminisme + unicité par assertion.

**Split PAR FOND (graine 42)** — 36 fonds train (108 paires) / **6 fonds eval (18 paires)** : l'éval porte sur des sujets **jamais vus** — ni leur choisi, ni aucun rejet. L'accuracy eval mesure la généralisation du style à des fonds nouveaux.

**SFT de référence (cellule 14)** — entraîné sur les choisis des fonds **train uniquement** (36 exemples, 3 epochs) au lieu de 5 exemples inline : les fonds d'eval restent totalement inconnus avant l'évaluation.

**DPO instrumenté (cellule 18)** — mini-batches de 4 (81 steps), courbe matplotlib de la perte par step + moyenne mobile, et à **chaque epoch** l'accuracy sur le jeu train ET sur les 18 paires d'eval disjointes — le contraste se construit sous les yeux de l'étudiant.

**Verdict à deux métriques (cellule 20)** — générations SFT vs DPO (2 fonds train + 1 fond eval jamais vu) + le tableau final : **accuracy TRAIN (mémorisation possible) vs accuracy EVAL disjointe (le vrai score)**, détail par famille de rejet, marge moyenne sur l'eval.

## Mesures (exécution GPU réelle — RTX 3070, kernel ft03-gpu)

**Reward proxy du modèle de base (cellule 10)** — préfère `chosen` sur **30/126 (24 %)** seulement, et le détail par famille est la vraie découverte :

| Famille de rejet | Base préfère chosen |
|---|---|
| vs vague | **1/42 (2 %)** |
| vs expéditive | 3/42 (7 %) |
| vs hors-sujet | 26/42 (62 %) |

Le proxy de log-prob **préfère le remplissage lisse** (vague = texte prévisible = haute log-prob) : c'est la pathologie du reward naïf, écrite en cellule 11.

**DPO 3 epochs (108 paires train, batch 4, beta 0,1, lr 1e-5)** :

| Epoch | Loss | Acc train | Acc EVAL disjointe |
|---|---|---|---|
| 1 | 0,655 | 94 % | **39 %** (sous le hasard) |
| 2 | 0,339 | 94 % | 56 % |
| 3 | 0,237 | 98 % | 56 % |

**Verdict final (cellule 20)** — train **106/108 (98 %)** vs eval disjointe **10/18 (56 %)**. Par famille sur l'eval : expéditive 5/6 (83 % — transfert net depuis 7 % au base), vague 2/6 (33 % — amélioré depuis 2 % mais loin du compte), hors-sujet 3/6 (50 %). Marge moyenne eval **+0,178 log-prob/token**. Générations SFT et DPO toutes deux semi-charabia : **DPO aligne un classement, il ne crée pas de compétence** — lecture écrite en cellule 21.

## Preuve d'exécution

```
PAPERMILL_EXIT=0  (kernel ft03-gpu, torch 2.13.0+cu126, RTX 3070 Laptop 8 Gio)
12/12 cellules code exécutées, 0 erreur, 2 figures embarquées (courbe perte DPO + moyennes)
validate_pr_notebooks.py origin/main <nb> : 1/1 PASS (12 cells)
detect_code_in_markdown_cells.py --check --baseline : OK, no new violations
pre-commit H.3 : PASS (hooks verts après re-stage fix-hr-separator)
```

## Acceptance FT-04 (#12428)

| Critère | État |
|---|---|
| ≥ 100 paires de préférences | **126** (assertion in-notebook) |
| Éval disjointe rapportée | 18 paires / 6 fonds jamais vus, accuracy par epoch + verdict final |
| Verdict sur l'écart métrique/générations | cellule 21 : train 98 % vs eval 56 % + générations inchangées commentées |
| Courbe de loss | matplotlib cellule 18 (perte DPO par step + moyenne mobile) |
| Zéro charabia non commenté | chaque sortie dégradée porte son verdict |
| Re-exécution C.2 | papermill end-to-end, outputs committés |

FT-03 (#12594) + cette PR couvrent les deux acceptance checklists de #12428. **Fermeture** : la dernière des deux PRs à merger peut porter `Closes #12428` (éditer le body avant merge) — je ne le mets pas ici pour éviter une fermeture automatique prématurée si #12594 est encore en review quand celle-ci passe.

See #12428

### CODE REVIEW

**Review by jsboige — COMMENTED** [review]
**[Hermes] Review FT-04 dataset 126 paires + eval disjointe (head 4069a4c, première review sur ce SHA).**

**Verdict : solide, prêt à merger — le message pédagogique (mémorisation vs généralisation) est exactement ce que l'issue #12428 demandait de rendre visible.**

Vérifications effectuées :

1. **Cohérence arithmétique du builder** — 42 fonds (20 définitions + 12 comparaisons + 10 pourquois) × 3 familles de rejets = 126 paires ✓. Split par fond : 36 fonds train / 6 fonds eval = 108/18 paires ✓. Les asserts `len == 126`, unicité `(prompt, famille) == 126`, et re-run `seed=42` identique sont dans le code, pas juste annoncés.

2. **Outputs authentiques** — timestamps d'exécution 2026-08-23T13:05-13:06Z (exécution du jour, cohérente avec le push), PyTorch 2.13.0+cu126 réel dans le boot log. Les trois familles de rejets (vague/expéditive/hors-sujet) apparaissent dans les générations montrées.

3. **La métrique honnête tient** — base model 30/126 (24%, classe à l'envers) → DPO epoch 1-3 : train 94/94/98% **vs eval disjoint 39/56/56%**. La phrase « à peine au-dessus du hasard (50 %) » est correcte et courageuse : 10/18 n'est pas significativement > 50% (binomiale p≈0.24). Le remplacement du « 100 % de préférences correctes » de l'ancienne version (6 paires inline) par cette lecture à deux colonnes est l'exacte réponse au problème de l'issue.

4. **Split par fond > split par paire** — bon choix méthodologique documenté dans le notebook : les 18 paires eval portent sur des sujets dont ni le chosen ni aucun rejet n'ont été vus (l'hors-sujet d'un fond eval utilise un chosen d'un autre fond eval — vérifié dans la construction).

Note CI (non bloquante) : les 2 « render volume delta » signalés correspondent aux figures matplotlib remplacées (courbes train/eval par epoch) — nouvelles images présentes (2 `image/png` ajoutées), pas de perte de contenu.

Rien à changer de mon côté.

**Comment by jsboige** [comment]
Rouge navlinks = defaut PRE-EXISTANT sur main, non reparable par cette lane (justification ecrite, cycle c.67) :

Les liens casses vises vivent dans `GameTheory-03d-Plan-de-deformation.ipynb` cell0 L3, qui reference `GameTheory-3b-Chambres-et-Murs.ipynb` et `GameTheory-3c-Le-Joueur-LLM.ipynb` — **inexistants** (les fichiers sont paddes `03b`/`03c` depuis le renommage zero-pad). Preuve locale firsthand sur le main actuel : les deux cibles `os.path.exists = False`, aucune cellule de cette PR ne touche un fichier GameTheory.

Le fix canonique est le rider **#12619** (po-2025:CoursIA-2, reconnu MERGE FIRST par po-2023 18:06Z) — il appartient a une autre lane et doit merger avant tout verdict vert ici. Le garde compte ces liens comme NEW car la baseline predates le renommage merge sur main ; la rupture vient de main, pas de cette PR.

Une fois #12619 merge : le sweep fast-lane (ombre) rejouera et passera (ou `gh run rerun <navlink-run> --failed` sur demande). Aucun geste lane possible d'ici la — `--ignore-red` documente.

**Comment by jsboige** [comment]
Constat navlink : le rouge check-navlinks de cette PR est un FANTOME mesure par ai-01 (2026-08-24T00:2xZ, DM lane) — merge-ref anterieure a b01ca3e3e (#12619), defaut deja corrige sur main. Pas d'update-branch pour l'effacer (file CI 1046 runs, ~97 workflows/tete). Cf #12567. Si un vrai travail repousse la branche, le fantome s'effacera dans la meme tete.

**Comment by jsboige** [comment]
Rouge PR gate = stale-rollup defect #10435, NON réparable par cette lane — famine CI #11860 confirmée.

État au cycle c.1331p447 :
- 78 checks SUCCESS (dont `Detect notebook changes` 05:23:31Z, `Golden-set execution` 05:25:40Z, `Papermill ratchet`, `validate-notebooks`, `Gitleaks`, `CodeQL`, `Fast lane`, `G-VAR-2/3`, `Require Grain tag`, `Require prev: non-closing`, `Require genre diversity vs prev`).
- `gh run rerun 32678727793` posté 08:06:07Z → toujours QUEUED après 5 min. File CI saturée (~47 in-flight, 30 queued). C'est la famine documentee #11860.
- `pr-gate-stale-sweep.yml` (cron 7,27,47) rejouera l'agrégation à 08:27Z (~16 min).

Ce commentaire vaut justification écrite pour picker `--ignore-red` (cf c.1331p442) : la cause est infra-only, le code est sain, le grain substance (FT-04 DPO tranche 2/2 #12428) est prêt à merger dès que la famine se résorbe.

**Comment by jsboige** [comment]
**Update 08:38Z** — empty-commit force-push tenté (`a7df23fdcc6aa82abf0cc547dd6012cac40abf5b` sur `fix/ft04-dataset-12428`) pour re-trigger fresh head CI. Nouveau PR gate run re-queued 08:37:54Z — toujours `QUEUED` après ~1 min, queue saturée ~100 in-flight (cf API `runs?status=queued`).

**Conclusion du cycle c.1331p447** : rouge PR gate = stale-rollup #10435 confirmé + famine CI #11860 confirmée, deux causes non réparables par cette lane. Justification écrite pour picker `--ignore-red` (cf c.1331p442) ; PR sera mergeable quand la famine se résorbe (prochain sweep 7,27,47 ≈ 08:47Z).
